packages/main/src/Menu.js

```diff
@@ -0,0 +1,490 @@
+		return this.effectiveDir === "rtl";
+	}
+
+	get placementType() {
+		const placement = this.isRtl ? "Left" : "Right";
+		return this._isSubMenu ? placement : "Bottom";
+	}
+
+	get verticalAlign() {
+		return this._isSubMenu ? "Top" : "Bottom";
+	}
+
+	get labelBack() {
+		return Menu.i18nBundle.getText(MENU_BACK_BUTTON_ARIA_LABEL);
+	}
+
+	get labelClose() {
+		return Menu.i18nBundle.getText(MENU_CLOSE_BUTTON_ARIA_LABEL);
+	}
+
+	get isPhone() {
+		return isPhone();
+	}
+
+	get isSubMenuOpened() {
+		return !!this._parentMenuItem;
+	}
+
+	get menuTitlePhone() {
+		return this.isSubMenuOpened ? this._parentMenuItem.text : this.title;
+	}
+
+	onBeforeRendering() {
+		!isPhone() && this._prepareCurrentItems(this.items);
+
+		const itemsWithChildren = this.itemsWithChildren;
+		const itemsWithIcon = this.itemsWithIcon;
+
+		this._currentItems.forEach(item => {
+			item.item._siblingsWithChildren = itemsWithChildren;
+			item.item._siblingsWithIcon = itemsWithIcon;
+		});
+	}
+
+	/**
+	 * Shows the Menu near the opener element.
+	 * @param {HTMLElement} opener the element that the popover is shown at
+	 * @public
+	 */
+	async showAt(opener) {
+		if (isPhone()) {
+			this._prepareCurrentItems(this.items);
+			this._parentItemsStack = [];
+		}
+		if (!this._isSubMenu) {
+			this._parentMenuItem = undefined;
+		}
+		(await this._getPopover()).showAt(opener);
+	}
+
+	/**
+	 * Closes the Menu.
+	 * @public
+	 */
+	close() {
+		if (Object.keys(this._popover).length) {
+			if (isPhone()) {
+				this._parentItemsStack = [];
+			}
+			this._popover.close();
+			this._popover.resetFocus();
+		}
+	}
+
+	async _getPopover() {
+		const id = this._id;
+		this._popover = (await this.getStaticAreaItemDomRef()).querySelector("[ui5-responsive-popover]");
+		this._popover.initialFocus = `${id}-menu-item-0`;
+		return this._popover;
+	}
+
+	_beforePopoverClose() {
+		if (Object.keys(this._openedSubMenuItem).length) {
+			this._openedSubMenuItem._preventSubMenuClose = false;
+			this._closeItemSubMenu(this._openedSubMenuItem);
+		}
+	}
+
+	_navigateBack(event) {
+		const parentMenuItem = this._parentItemsStack.pop();
+		this.focus();
+		if (parentMenuItem) {
+			this._prepareCurrentItems(parentMenuItem.parentElement.items);
+			this._parentMenuItem = this._parentItemsStack.length ? this._parentItemsStack[this._parentItemsStack.length - 1] : undefined;
+		}
+	}
+
+	_prepareCurrentItems(items) {
+		this._currentItems = items.map((item, index) => {
+			return {
+				item,
+				position: index + 1,
+				ariaHasPopup: item.hasChildren ? "menu" : undefined,
+			};
+		});
+	}
+
+	_createSubMenu(item, openerId) {
+		const subMenu = document.createElement(this.constructor.getMetadata().getTag());
+
+		subMenu._isSubMenu = true;
+		subMenu.setAttribute("id", `submenu-${openerId}`);
+		subMenu._parentMenuItem = item;
+		const subItems = item.children;
+		let clonedItem,
+			idx;
+		for (idx = 0; idx < subItems.length; idx++) {
+			clonedItem = subItems[idx].cloneNode(true);
```

Comment:
We don't expect to have too much sub items, but I suggest using the Document Fragment as touching the DOM comes at a cost. If we are appending many elements to the DOM, it's better t to append them all at once, rather than one at a time. The document fragment is in memory and not part of the main DOM tree, appending children to it does not cause page reflow. Basically you will keep the same syntax and a slight change.
Example:
https://developer.mozilla.org/en-US/docs/Web/API/Document/createDocumentFragment#javascript

Reply:
Done